File: next-frontend/lib/types/deep-memory.ts
export interface GenerateResponse {
  job_id: string;
  training_run_id: string;
  total_chunks: number;
  message: string;
}

export interface TrainResponse {
  job_id: string;
  training_run_id: string;
  message: string;
}

export interface TrainingRunSummary {
  id: string;
  status: string;
  pair_count: number;
  processed_chunks: number;
  total_chunks: number;
  error_message: string | null;
  metrics: Record<string, number>;
  started_at: string;
  completed_at: string | null;
}

export interface SamplePair {
  question_text: string;
  chunk_preview: string;
  relevance_score: number;
}

export interface TrainingRunDetail extends TrainingRunSummary {
  total_chunks: number;
  processed_chunks: number;
  deeplake_job_id: string | null;
  error_message: string | null;
  sample_pairs: SamplePair[];
  statistics: {
    avg_questions_per_chunk: number;
    chunk_coverage_pct: number;
  };
}

export interface ProceedResponse {
  job_id: string;
  training_run_id: string;
  message: string;
}

export interface DeepMemorySettings {
  enabled: boolean;
  last_trained_at: string | null;
  last_training_run_id: string | null;
  can_enable: boolean;
  total_chunks: number;
  trained_chunk_count: number;
  has_blocking_run: boolean;
  blocking_run_id: string | null;
  blocking_run_status: string | null;
  is_cloud: boolean;
}

export interface DeepMemoryJobUpdate {
  id: string;
  status: string;
  progress: number;
  processed_chunks?: number;
  total_chunks?: number;
  pair_count?: number;
  message?: string;
  metrics?: Record<string, number>;
  error_message?: string;
}

File: backend/app/models/deep_memory.py
from datetime import datetime

from pydantic import BaseModel


class GenerateRequest(BaseModel):
    pass


class GenerateResponse(BaseModel):
    job_id: str
    training_run_id: str
    total_chunks: int
    message: str


class TrainRequest(BaseModel):
    training_run_id: str


class TrainResponse(BaseModel):
    job_id: str
    training_run_id: str
    message: str


class SamplePair(BaseModel):
    question_text: str
    chunk_preview: str
    relevance_score: float


class TrainingRunSummary(BaseModel):
    id: str
    status: str
    pair_count: int
    processed_chunks: int
    total_chunks: int
    error_message: str | None = None
    metrics: dict
    started_at: datetime
    completed_at: datetime | None


class TrainingRunDetail(TrainingRunSummary):
    total_chunks: int
    processed_chunks: int
    deeplake_job_id: str | None
    error_message: str | None
    sample_pairs: list[SamplePair]
    statistics: dict


class TrainingRunListResponse(BaseModel):
    runs: list[TrainingRunSummary]


class DeepMemorySettingsResponse(BaseModel):
    enabled: bool
    last_trained_at: datetime | None
    last_training_run_id: str | None
    can_enable: bool
    total_chunks: int
    trained_chunk_count: int
    has_blocking_run: bool = False
    blocking_run_id: str | None = None
    blocking_run_status: str | None = None
    is_cloud: bool = True


class ProceedRequest(BaseModel):
    training_run_id: str


class ProceedResponse(BaseModel):
    job_id: str
    training_run_id: str
    message: str


class UpdateSettingsRequest(BaseModel):
    enabled: bool
